### ISSUE TITLE
"errSecInternalComponent" error during the test.txt signing phase

### ISSUE THREAD

**Issue by YasharAzari** [danielpaulus/app-signer#1]
Hello daniel !

I was trying to use the app-signer out of curiosity, just to check how it works as it seems to be a really nice alternative to other toolings.

While trying it with a test ipa, i always met this error during the .txt signing test phase: `error="exit status 1" output="Warning: unable to build chain to self-signed root for signer \"iPhone Developer: ################## (############)\"\n/var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/sign/test.txt: errSecInternalComponent\n"`

For troubleshooting i tried to:
- check the permission of the .txt file
- check the permission of the temporary keychain created
- run the command as sudo
- run the signing command independently by keeping a copy of the final keychain generated
- check the content of the temporary keychain itself within the keychain access.app (seems to be always empty without any certificate inside. Not sure if it's expected)
- changed the code so that it builds and created a signing workspace outside `/var`
- Used it on two different mac machine running macOS 12.1, one with a M1 cpu, the other an intel one

None of the solutions worked so far so i'm probably sure i must have missed something along the way :(

Searching this error on the web, it mostly comes up when users are codesigning with the Default Keychain (login) which is password protected. It seems weird that it comes up here as we are using our own keychain.

Here is the stacktrace for reference:

```
INFO[0000] starting iOS appsigner                        args="[./app-signer --udid=<device_id> --p12password=<pwd> --profilespath=<path_to_folder_with_provisioning_file+p12> --ipa=<path_to_ipa> --output=<output_folder>]"
INFO[0000] cleaning workdir
INFO[0000] parsing profile '<path_to_folder_with_provisioning_file+p12>/########.mobileprovision'
INFO[0000] found 1 profiles
INFO[0000] extracting files for profile: ########
INFO[0000] extracting entitlements to: '/var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/########-entitlements.plist'
INFO[0000] extracting signing certificate ############################# to: '/var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/########-signingcert.p12'
INFO[0000] keychain created: /var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/appsigner.keychain
INFO[0000] installing /var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/########-signingcert.p12 to keychain
INFO[0000] codesign test signing failed                  cert=############################# cmd="/usr/bin/codesign -vv --keychain /var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/appsigner.keychain --deep --force --sign ####################################### /var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/sign/test.txt" error="exit status 1" output="Warning: unable to build chain to self-signed root for signer \"iPhone Developer: ################## (############)\"\n/var/folders/hb/ct389svx7yv43p_3s59mm4mw0000gn/T/pattern1559172448/sign/test.txt: errSecInternalComponent\n"
ERRO[0000] test signing failedexit status 1
```
--------
The other question i had is more related to the code itself:
So far it seems we need to provide a reference device ID, even if our mobileprovision file was generated with multiple devices.
It seems to be related to this section at `api/workspace.go:38` where we expect a folder with a list of .mobileprovision files instead of just one:
- As and output, we have a table of profiles. 
- The udid is then only used as a reference to select the correct mobileprovision data here : `codesign/profileparser.go:54`. I'm not sure if this is the correct interpretation though.

I was wondering if it was something mandatory and if it's possible to expose another function to resign without a device id as reference ? (instead we can give the path to a single mobileprovision and p12 or maybe sign the same ipa with all the mobileprovision file available in the folder).

**Comment by YasharAzari** [comment]
I was able to solve the test signing issue after more investigation.
To give some context: I was working with a fresh install of macos, with xcode installed but never used, with a certificate generated the same day or day prior and never installed in the system keychain.

TLDR:
If codesign does not work in your terminal and you see this warning: `Warning: unable to build chain to self-signed root for signer`, try to sign something (a project for instance) with the certificate in the global keychain using xcode (or just open a project within xcode), then check if it works or not. 
<img width="200" alt="Screenshot 2022-01-27 at 16 05 24" src="https://user-images.githubusercontent.com/20423111/151385514-ca90ed0a-0786-4a0d-b060-d9eb684164aa.png">
Verify that your certificate is valid (or became valid) in your keychain.
If successful, codesign again within your terminal and it should work.

---
Troubleshooting steps:

1) I first installed the certificate directly on the system keychain and tried to do a typical codesign without the temporary keychain created by app-signer. It failed with the same error message and warning
2) Looking at the keychain, while the certificate was there, it was not verified/trusted and had a red cross.

3) Then i tried to open a project and try to build and sign something with xcode for the first time on the machine, just to check what would happen.

I noticed two things:
- xcode did an export command before codesigning something: `export CODESIGN_ALLOCATE=/Applications/<your_xcode>.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate;`. It may have been something important, but doesn't seem to be the case.
- the second thing was the certificate itself became trusted after opening xcode. I stopped the process at this point, and within the keychain the certificate became trusted and codesign started working again when using the terminal.

I suppose something had to be initialised before, and using xcode helped with that. It's still a bit mysterious :)

4) After these steps, i ran app-signer again and it worked flawlessly by using the certificate in the global keychain (as expected i was prompted with the password input popup)

5) I deleted everything in the global keychain and ran app-signer again without issues

---
Additional informations:

As seen in many stackoverflow and apple forum issues, if you ssh to your machine, codesign and see `errSecInternalComponent` when trying to sign something with a certificate in your system keychain:
```
./testfile.txt: replacing existing signature
./testfile.txt: errSecInternalComponent
```

Make sure the keychain containing your certificate is unlocked. If it's not, you can run `security unlock-keychain`, `security unlock-keychain login.keychain` or `security unlock-keychain <your_custom_keychain>.keychain`

Nevertheless this issue should probably be non-existent if you are using tools using the same logic as app-signer
